src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -505,56 +526,42 @@ private void onMqttSubscribe(
                     }
                 }
 
-                for (int progress = offset; progress < limit; progress = subscription.limit())
+                for (int progress = offset, ackIndex = 0; progress < limit; progress = mqttSubscription.limit(), ackIndex++)
                 {
-                    subscription = mqttSubscriptionRO.tryWrap(buffer, progress, limit);
-                    if (subscription == null)
+                    mqttSubscription = mqttSubscriptionRO.tryWrap(buffer, progress, limit);
+                    if (mqttSubscription == null)
                     {
                         break;
                     }
-
-                    int reasonCode = 0x8f;
-                    final String topicFilter = subscription.topicFilter().asString();
+                    final String topicFilter = mqttSubscription.topicFilter().asString();
                     final RouteFW route = resolveTarget(routeId, authorization, topicFilter);
 
                     if (route != null)
                     {
-                        final MqttServerStream serverStream = new MqttServerStream();
-
                         final long newRouteId = route.correlationId();
                         final long newInitialId = supplyInitialId.applyAsLong(newRouteId);
                         final long newReplyId = supplyReplyId.applyAsLong(newInitialId);
 
                         final MessageConsumer newTarget = router.supplyReceiver(newInitialId);
 
-                        final MqttBeginExFW beginEx = mqttBeginExRW
-                            .wrap(writeBuffer, BeginFW.FIELD_OFFSET_EXTENSION, writeBuffer.capacity())
-                            .typeId(mqttTypeId)
-                            .role(r -> r.set(MqttRole.RECEIVER))
-                            .clientId("client")
-                            .topic(topicFilter)
-                            .subscriptionId(subscriptionId)
-                            .build();
+                        final MqttServerStream serverStream = new MqttServerStream(newTarget,
+                            newRouteId, newInitialId, newReplyId, ackIndex, packetId, subscription);
 
-                        serverStream.doMqttBeginEx(newTarget, newRouteId, newInitialId,
-                            decodeTraceId, beginEx.buffer(), beginEx.offset(), beginEx.sizeof());
+                        serverStream.doMqttBeginEx(decodeTraceId, topicFilter, subscriptionId);
 
                         correlations.put(newReplyId, serverStream);
 
-                        reasonCode = 0x00;
-
-                        mqttStreams.put(topicFilter, serverStream);
+                        subscribers.put(topicFilter, serverStream);
+                        reasonCodeCount++;
```

Comment:
This increments `reasonCodeCount` only when routeable, but we want the total count.
Can we just use `ackIndex` after the loop has completed? Perhaps define `ackIndex` outside the loop, but still increment it each iteration.
Perhaps rename `ackIndex` to `ackCount` since the count will be increasing, and remove `reasonCodeCount`.

Reply:
In fact, we could use `subscription.ackCount` directly instead.